[JSC] Refactor DFG node child iteration macros and unify ForwardingFunc usage

hyjorc1 opened the pull request:
#### c149310ffe3055d286df2c60f4fc46a92348eede
<pre>
[JSC] Refactor DFG node child iteration macros and unify ForwardingFunc usage
<a href="https://bugs.webkit.org/show_bug.cgi?id=290954">https://bugs.webkit.org/show_bug.cgi?id=290954</a>
<a href="https://rdar.apple.com/148468379">rdar://148468379</a>

Reviewed by Keith Miller.

The motivation for this patch is that when cloning DFG nodes with VarArgs,
we should also copy the empty edges. However, doToChildren* only iterate
non-empty edges. To support this, the patch makes the following changes:

1. Consolidates DFG_NODE_DO_TO_CHILDREN and related macros into a generic
   macro with customizable child filtering and loop behavior.
2. Introduces ForwardingFunc as a reusable template class to eliminate repeated
   class definitions across doToChildren*, reducing boilerplate.
3. Updates doToChildren, doToAllChildren, and doToChildrenBreakOnDone to
   use the shared ForwardingFunc.
4. Replaces the custom child iteration logic in cloneNodeImpl with doToAllChildren
   for improved clarity and consistency.

* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp:
(JSC::DFG::CloneHelper::cloneNodeImpl):
* Source/JavaScriptCore/dfg/DFGGraph.h:

Canonical link: <a href="https://commits.webkit.org/293197@main">https://commits.webkit.org/293197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7c0dfc7891a8d144a6c4ee9c2df76c84c55004d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98164 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8022 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103281 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48693 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74737 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31909 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101168 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13706 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88692 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55097 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6633 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48135 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/90845 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83464 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105657 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/96790 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25250 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83721 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25623 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84877 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83176 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21014 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27820 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5503 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18890 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25209 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120411 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25029 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33733 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28345 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26604 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -778,59 +788,43 @@ class Graph final : public virtual Scannable {
+    ALWAYS_INLINE void doToChildren(Node* node, const ChildFunctor& functor)
     {
-        DFG_NODE_DO_TO_CHILDREN(*this, node, functor);
+        DFG_NODE_DO_TO_CHILDREN(*this, node, (ForwardingFunc<ChildFunctor>(functor)));
```
Do you need the `ChildFunctor` here?

hyjorc1 replied:
It seems armv7 complains about the missing template arg. Discussed with Keith offline. So, just use the current approach.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -778,59 +788,43 @@ class Graph final : public virtual Scannable {
+    };
+
     template<typename ChildFunctor>
-    ALWAYS_INLINE void doToChildrenWithNode(Node* node, const ChildFunctor& functor)
+    ALWAYS_INLINE void doToChildren(Node* node, const ChildFunctor& functor)
```
I think this can be `ALWAYS_INLINE void doToChildren(Node* node, const Invocable<IterationStatus(Node*)> auto& functor)`

hyjorc1 replied:
It seems armv7 complains about the missing template arg. Discussed with Keith offline. So, just use the current approach.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGGraph.h:
```diff
@@ -778,59 +788,43 @@ class Graph final : public virtual Scannable {
 
     template<typename ChildFunctor>
-    ALWAYS_INLINE void doToChildrenWithCheck(Node* node, const ChildFunctor& functor)
+    ALWAYS_INLINE void doToChildrenBreakOnDone(Node* node, const ChildFunctor& functor)
```
Ditto.

hyjorc1 replied:
It seems armv7 complains about the missing template arg. Discussed with Keith offline. So, just use the current approach.

hyjorc1 committed:
[JSC] Refactor DFG node child iteration macros and unify ForwardingFunc usage
https://bugs.webkit.org/show_bug.cgi?id=290954
rdar://148468379

Reviewed by Keith Miller.

The motivation for this patch is that when cloning DFG nodes with VarArgs,
we should also copy the empty edges. However, doToChildren* only iterate
non-empty edges. To support this, the patch makes the following changes:

1. Consolidates DFG_NODE_DO_TO_CHILDREN and related macros into a generic
   macro with customizable child filtering and loop behavior.
2. Introduces ForwardingFunc as a reusable template class to eliminate repeated
   class definitions across doToChildren*, reducing boilerplate.
3. Updates doToChildren, doToAllChildren, and doToChildrenBreakOnDone to
   use the shared ForwardingFunc.
4. Replaces the custom child iteration logic in cloneNodeImpl with doToAllChildren
   for improved clarity and consistency.

* Source/JavaScriptCore/dfg/DFGCloneHelper.cpp:
(JSC::DFG::CloneHelper::cloneNodeImpl):
* Source/JavaScriptCore/dfg/DFGGraph.h:

Canonical link: https://commits.webkit.org/293197@main
